components/tikv_alloc/src/trace.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! This module provides some utilities to define the tree hierarchy to trace memory.
+//!
+//! A memory trace is a tree that records how much memory its children and itself
+//! uses, It doesn't need to match any function stacktrace, instead it should
+//! have logically meaningful layout.
+//!
+//! For example, memory usage should be divided into several components under the
+//! root scope: TiDB EndPoint, Transaction, Raft, gRPC etc. TiDB EndPoint can divide
+//! its children by queries, while Raft can divide memory by store and apply. Name
+//! are defined as number for better performance. In practice, it can be mapped to
+//! enumerates instead.
+//!
+//! To define a memory trace tree, we can use the `mem_trace` macro. The `mem_trace`
+//! macro constructs every node as a `MemoryTraceNode` which implements `MemoryTrace` trait.
+//! We can also define a specified tree node by implementing `MemoryTrace` trait.
+
+use std::fmt::{self, Display};
+use std::num::NonZeroU64;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+type HashMap<K, V> =
+    std::collections::HashMap<K, V, std::hash::BuildHasherDefault<fxhash::FxHasher>>;
+
+#[derive(Debug, PartialEq, Copy, Clone, Hash, Eq)]
+pub enum Id {
+    Name(&'static str),
+    Number(u64),
+}
+
+impl Id {
+    pub fn readable_name(&self) -> String {
+        match self {
+            Id::Name(s) => {
+                let mut s = s.replace('_', "");
```

Comment:
Seems `_` is replaced, it conflicts with `split`? BTW it's better to add a test case for it.